docs/development/code-guideline/javascript.md

```diff
@@ -1,12 +1,1550 @@
+};
+```
+
+**Good:**
+
+```javascript
+const addItemToCart = (cart, item) => {
+  return [...cart, { item, date: Date.now() }];
+};
+```
+
+**[⬆ back to top](#table-of-contents)**
+
+### Don't write to global functions
+
+Polluting globals is a bad practice in JavaScript because you could clash with another
+library and the user of your API would be none-the-wiser until they get an
+exception in production. Let's think about an example: what if you wanted to
+extend JavaScript's native Array method to have a `diff` method that could
+show the difference between two arrays? You could write your new function
+to the `Array.prototype`, but it could clash with another library that tried
+to do the same thing. What if that other library was just using `diff` to find
+the difference between the first and last elements of an array? This is why it
+would be much better to just use ES2015/ES6 classes and simply extend the `Array` global.
+
+**Bad:**
+
+```javascript
+Array.prototype.diff = function diff(comparisonArray) {
+  const hash = new Set(comparisonArray);
+  return this.filter(elem => !hash.has(elem));
+};
+```
+
+**Good:**
+
+```javascript
+class SuperArray extends Array {
+  diff(comparisonArray) {
+    const hash = new Set(comparisonArray);
+    return this.filter(elem => !hash.has(elem));
+  }
+}
+```
+
+**[⬆ back to top](#table-of-contents)**
+
+### Favor functional programming over imperative programming
+
+JavaScript isn't a functional language in the way that Haskell is, but it has
+a functional flavor to it. Functional languages can be cleaner and easier to test.
+Favor this style of programming when you can.
+
+**Bad:**
+
+```javascript
+const programmerOutput = [
+  {
+    name: 'Uncle Bobby',
+    linesOfCode: 500
+  },
+  {
+    name: 'Suzie Q',
+    linesOfCode: 1500
+  },
+  {
+    name: 'Jimmy Gosling',
+    linesOfCode: 150
+  },
+  {
+    name: 'Gracie Hopper',
+    linesOfCode: 1000
+  }
+];
+
+let totalOutput = 0;
+
+for (let i = 0; i < programmerOutput.length; i++) {
+  totalOutput += programmerOutput[i].linesOfCode;
+}
+```
+
+**Good:**
+
+```javascript
+const programmerOutput = [
+  {
+    name: 'Uncle Bobby',
+    linesOfCode: 500
+  },
+  {
+    name: 'Suzie Q',
+    linesOfCode: 1500
+  },
+  {
+    name: 'Jimmy Gosling',
+    linesOfCode: 150
+  },
+  {
+    name: 'Gracie Hopper',
+    linesOfCode: 1000
+  }
+];
+
+const totalOutput = programmerOutput.reduce(
+  (totalLines, output) => totalLines + output.linesOfCode,
+  0
+);
+```
+
+**[⬆ back to top](#table-of-contents)**
+
+### Encapsulate conditionals
+
+**Bad:**
+
+```javascript
+if (fsm.state === 'fetching' && isEmpty(listNode)) {
+  // ...
+}
+```
+
+**Good:**
+
+```javascript
+function shouldShowSpinner(fsm, listNode) {
+  return fsm.state === 'fetching' && isEmpty(listNode);
+}
+
+if (shouldShowSpinner(fsmInstance, listNodeInstance)) {
+  // ...
+}
+```
+
+**[⬆ back to top](#table-of-contents)**
+
+### Avoid negative conditionals
+
+**Bad:**
+
```

Comment:
I'm missing an explanation for this, care to add one?

Reply:
Looking in to it. :)